Add an option to allow the user to change the full-text search behavior in the admin UI.

MikeAlhayek opened the pull request:
Fix #13886

jtkech commented on src/OrchardCore/OrchardCore.Contents.Core/ContentSearchOption.cs:
```diff
@@ -0,0 +1,46 @@
+
+namespace OrchardCore.Contents;
+
+public class ContentSearchOption
```
If we keep the options pattern, maybe `ContentsAdminListFilterOptions` ;)

Or `ContentsAdminListOptions`, at least `ContentsSearchOptions`, I don't know.

MikeAlhayek replied:
done

jtkech commented on src/OrchardCore/OrchardCore.Contents.Core/ContentSearchOption.cs:
```diff
@@ -0,0 +1,46 @@
+
+public class ContentSearchOption
+{
+    private readonly Dictionary<string, string> _maps = new(StringComparer.OrdinalIgnoreCase);
```
Do we really need to be case insensitive ;)

MikeAlhayek replied:
Yes because we'll read the selectedContentType from the UI "URL" or filter.

jtkech commented on src/OrchardCore/OrchardCore.Contents.Core/ContentSearchOption.cs:
```diff
@@ -0,0 +1,46 @@
+    {
+        ArgumentNullException.ThrowIfNull(contentType, nameof(contentType));
+
+        if (_maps.ContainsKey(contentType))
```
We don't need to check if it exists.

jtkech replied:
Maybe `RemoveDefaultTermName()`

jtkech commented on src/OrchardCore.Modules/OrchardCore.Contents/Services/DefaultContentsAdminListQueryService.cs:
```diff
@@ -1,45 +1,105 @@
 {
     public class DefaultContentsAdminListQueryService : IContentsAdminListQueryService
     {
+        public const string DefaultTermName = "text";
```
Maybe define this constant in the options class instead (if we keep the options pattern).

jtkech commented on src/OrchardCore/OrchardCore.Contents.Core/ContentSearchOption.cs:
```diff
@@ -0,0 +1,46 @@
+    /// </summary>
+    /// <param name="contentType"></param>
+    /// <param name="termName"></param>
+    public void AddTerm(string contentType, string termName)
```
Maybe `AddDefaultTermName()`

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Contents/Services/DefaultContentsAdminListQueryService.cs:
```diff
@@ -1,45 +1,105 @@
+
+            if (defaultTermNode is not null)
+            {
+                defaultTermName = GetDefaultTermName(selectedContentType);
```
Yes. I think we have to rely of the content type. We can't target custom index for any content type.

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Contents/Services/DefaultContentsAdminListQueryService.cs:
```diff
@@ -1,45 +1,105 @@
+            {
+                defaultTermName = GetDefaultTermName(selectedContentType);
+
+                if (defaultTermName != defaultTermNode.TermName || defaultOperator != defaultTermNode.Operation)
```
done.

jtkech commented on src/OrchardCore.Modules/OrchardCore.Contents/Services/DefaultContentsAdminListQueryService.cs:
```diff
@@ -15,31 +18,86 @@ public class DefaultContentsAdminListQueryService : IContentsAdminListQueryServi
             _session = session;
             _serviceProvider = serviceProvider;
             _contentsAdminListFilters = contentsAdminListFilters;
+            _contentSearchOptions = contentSearchOptions.Value;
```
Maybe change the var names too.

MikeAlhayek replied:
done

jtkech commented on src/OrchardCore.Modules/OrchardCore.Contents/Services/DefaultContentsAdminListQueryService.cs:
```diff
@@ -15,31 +18,86 @@ public class DefaultContentsAdminListQueryService : IContentsAdminListQueryServi
+            await _contentsAdminListFilters
+                .InvokeAsync((filter, model, query, updater) => filter.FilterAsync(model, query, updater), model, query, updater, _logger);
+
+            if (defaultTermName != defaultTermNode?.TermName || defaultOperator != defaultTermNode?.Operation)
```
Here too the operator checking seems useless.

Do we really need to restore the defaultTermNode?

MikeAlhayek replied:
Check the code again after adding the double quotes please.

jtkech commented on src/docs/releases/1.7.0.md:
````diff
@@ -21,10 +21,54 @@ The `Seo` feature now provides `robots.txt` out of the box when the filesystem d
+ ```C#
+services.Configure<ContentsAdminListFilterOptions>(options =>
+{
+    options.AddTerm("Product", "product-text");
````
`AddTerm()` => `AddDefaultTermName()`

jtkech replied:
Hmm, not sure we can use a dash `-` in the term name, just need to be tried.

jtkech commented on src/OrchardCore/OrchardCore.Contents.Core/ContentsAdminListFilterOptions.cs:
```diff
@@ -0,0 +1,45 @@
+    /// <returns></returns>
+    public bool TryGetValue(string contentType, out string termName)
+    {
+        return _maps.TryGetValue(contentType, out termName);
```
Maybe `TryGetValue()` => `TryGetDefaultTermNameByContentType()` or if too long `TryGetDefaultTermName()`

In case we may have other options in the future.

Hmm, then maybe `_maps` => `_defaultTermNamesByContentType` or just `_defaultTermNames`

jtkech commented:
I did another review

LGTM but maybe worth to be able to also define the default operator, hmm as I remember we also needed to define a full quoted string because when searching for `pete seeger` the customer only want results containing the whole, no result like `pete foo played music with albert seeger`.

But I'm not against if we agree that it could be done in separate PRs.

MikeAlhayek commented:
> I did another review
> 
> LGTM but maybe worth to be able to also define the default operator, hmm as I remember we also needed to define a full quoted string because when searching for `pete seeger` the customer only want results containing the whole, no result like `pete foo played music with albert seeger`.
> 
> But I'm not against if we agree that it could be done in separate PRs.

 made some more changes and added double-quotes. Let me know if you see anything else.

jtkech commented on src/OrchardCore.Modules/OrchardCore.Contents/Services/DefaultContentsAdminListQueryService.cs:
```diff
@@ -1,45 +1,128 @@
+
+                var selectedContentType = GetSelectedContentType(model);
+
+                if (selectedContentType != null)
```
Maybe `is not null`, and no blank line before, that's a new habit I have ;)

MikeAlhayek replied:
Why are we used `is not null` now instead of `!= null`

jtkech replied:
Not so recent but it's a new c# feature, belonging to pattern matching if I remember correctly

jtkech commented on src/OrchardCore.Modules/OrchardCore.Contents/Services/DefaultContentsAdminListQueryService.cs:
```diff
@@ -1,45 +1,128 @@
+            await _contentsAdminListFilters
+                .InvokeAsync((filter, model, query, updater) => filter.FilterAsync(model, query, updater), model, query, updater, _logger);
+
+            if (shouldRestoreDefault)
```
Do we really need to restore the default?

MikeAlhayek replied:
I think we do since we swap `text` with a custom value. But not sure other you had it done in your version :)

jtkech replied:
> But not sure other you had it done in your version :)

Oups ;) Hmm I think we don't need it now that we are using options, worth to try

MikeAlhayek replied:
I think we need it to remove the added double quotes. But something isn't right because when we add the quote using the UnaryNode it does not change the behavior search. I'll look more at it tomorrow

jtkech replied:
Just tried, yes we need to remove the added quotes otherwise they are then rendered.

I just tried the quoted string option (without any default term name per content type), it works on my side.

MikeAlhayek replied:
@jtkech 
Feel free to commit any changes you see fit. Maybe I am missing the idea behind the double quotes.

jtkech commented on src/OrchardCore.Modules/OrchardCore.Contents/Services/DefaultContentsAdminListQueryService.cs:
```diff
@@ -1,45 +1,128 @@
+            {
+                var typeTermNode = model.FilterResult.OfType<ContentTypeFilterNode>().FirstOrDefault();
+
+                if (typeTermNode != null)
```
Maybe `is not null`, and no blank line before.

MikeAlhayek replied:
done.

jtkech commented on src/OrchardCore.Modules/OrchardCore.Contents/Services/DefaultContentsAdminListQueryService.cs:
```diff
@@ -1,45 +1,128 @@
+
+                var sterotypeTermNode = model.FilterResult.OfType<StereotypeFilterNode>().FirstOrDefault();
+
+                if (sterotypeTermNode != null)
```
idem

jtkech commented on src/docs/releases/1.7.0.md:
````diff
@@ -21,10 +21,54 @@ The `Seo` feature now provides `robots.txt` out of the box when the filesystem d
+ ```C#
+services.Configure<ContentsAdminListFilterOptions>(options =>
+{
+    options.AddTerm("Product", "producttext");
````
`AddDefaultTermName`

MikeAlhayek replied:
done.

jtkech commented:
Okay, let @sebastienros decide which options are worth to support, maybe also the default operation, and the `Full Quoted String` as a separate option with an higher precedence. Yes I think that if we support the `Quoted String` it should be an option.

Maybe the options would need to be able to return a full `DefaultTermNode`.

In my case I was checking if there is no explicit operator in use, maybe we don't need this, just take into account the current options, but not sure, would need to focus on it more.

MikeAlhayek commented:
@jtkech I added `UseQuotationMarks` option which is false by default so we don't change existing behavior. Currently, the user can achieve this by wrapping the string with double-quotes themself. This is better I think because the user has full control. The user should only turn on this option if they really need to always use double quote.

jtkech commented:
> Currently, the user can achieve this by wrapping the string with double-quotes

Does it work? As I remember it was not working when I tried it, but not sure now

I don't like the `UseQuotationMarks` option name, maybe `AsQuotedString`

jtkech commented on src/OrchardCore/OrchardCore.Contents.Core/ContentsAdminListFilterOptions.cs:
```diff
@@ -0,0 +1,51 @@
+
+    /// <summary>
+    /// When true, the entire searched terms will be enclose within quotation marks, it creates an exact match search.
+    /// This means that search engines will treat the entire term as a single keyword or phrase, rather than individual words.
```
`enclose` => `enclosed`

I would say that the `whole text is treated as a single term`

jtkech commented on src/OrchardCore/OrchardCore.Contents.Core/ContentsAdminListFilterOptions.cs:
```diff
@@ -0,0 +1,51 @@
+    public bool UseQuotationMarks { get; set; }
+
+    /// <summary>
+    /// Gets the term-name to use by the given content type.
```
`term-name` => `term name` as in other places, maybe `default term name`

MikeAlhayek replied:
done

jtkech commented on src/OrchardCore/OrchardCore.Contents.Core/ContentsAdminListFilterOptions.cs:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /// <summary>
+    /// Removes the content type from from the options if exists.
```
`from from` => `from`
`if exists` => `if it exists`

Maybe `Removes the default term name for the provided content type`

MikeAlhayek replied:
done

jtkech commented on src/OrchardCore/OrchardCore.Contents.Core/ContentsAdminListFilterOptions.cs:
```diff
@@ -0,0 +1,51 @@
+    /// <returns></returns>
+    public bool TryGetDefaultTermName(string contentType, out string termName)
+    {
+        return _defaultTermNames.TryGetValue(contentType, out termName);
```
Here you can use an expression-bodied `=> _defaultTermNames.TryGetValue(contentType, out termName);`

jtkech commented on src/OrchardCore/OrchardCore.Contents.Core/ContentsAdminListFilterOptions.cs:
```diff
@@ -0,0 +1,51 @@
+    /// <param name="termName"></param>
+    public void AddDefaultTermName(string contentType, string termName)
+    {
+        ArgumentNullException.ThrowIfNull(contentType, nameof(contentType));
```
Not sure it's useful to add a null check here.

MikeAlhayek replied:
removed.

MikeAlhayek commented:
@jtkech I renamed `UseQuotationMarks` to `UseExactMatch`. I cleaned up the doc blocks and cleaned up the code. I also tested with and without `UseExactMatch` and it is working as expected. By default `UseExactMatch` is set to `false` so we do not change the default behavior of Orchard.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Contents/Services/DefaultContentsAdminListQueryService.cs:
```diff
@@ -1,45 +1,124 @@
 {
     public class DefaultContentsAdminListQueryService : IContentsAdminListQueryService
     {
+        private readonly static List<string> _operators = new() { "OR", "AND", "||", "&&" };
```
string[]

jtkech commented:
Okay, I will look at it this night, I may do some last minor changes and then approve (or merge).

jtkech commented:
Okay, I did some last changes and approved, I let you review my changes
